// @pothos/plugin-prisma/esm/util/datamodel.js
import { PrismaObjectRef } from '../object-ref.js';
import { getDMMF } from './get-client.js';
export const refMap = new WeakMap();
export const findUniqueMap = new WeakMap();
export const includeForRefMap = new WeakMap();
export function getRefFromModel(name, builder) {
    if (!refMap.has(builder)) {
        refMap.set(builder, new Map());
    }
    const cache = refMap.get(builder);
    if (!cache.has(name)) {
        cache.set(name, new PrismaObjectRef(name));
    }
    return cache.get(name);
}
export function getRelation(name, builder, relation) {
    const fieldData = getFieldData(name, builder, relation);
    if (fieldData.kind !== "object") {
        throw new Error(`Field ${relation} of model '${name}' is not a relation (${fieldData.kind})`);
    }
    return fieldData;
}
export function getFieldData(name, builder, fieldName) {
    const modelData = getModel(name, builder);
    const fieldData = modelData.fields.find((field) => field.name === fieldName);
    if (!fieldData) {
        throw new Error(`Field '${fieldName}' not found in model '${name}'`);
    }
    return fieldData;
}
export function getModel(name, builder) {
    const dmmf = getDMMF(builder);
    const modelData = dmmf.datamodel.models.find((model) => model.name === name);
    if (!modelData) {
        throw new Error(`Model '${name}' not found in DMMF`);
    }
    return modelData;
}
export function getDelegateFromModel(client, model) {
    const lowerCase = `${model.slice(0, 1).toLowerCase()}${model.slice(1)}`;
    const delegate = lowerCase in client ? client[lowerCase] : null;
    if (!delegate) {
        throw new Error(`Unable to find delegate for model ${model}`);
    }
    return delegate;
}
